cm_ovirt_vm_creator.py

```diff
@@ -143,47 +143,103 @@ def construct_search_by_name_query(node_name):
-                vm = vm_service.get()
-                print("vm.status = %s, vm.fqdn= '%s'" % (vm.status, vm.fqdn), file=sys.stderr)
-                if vm.status == types.VmStatus.UP:
-                    break
-                counter += 1
-
-            if vm.status != types.VmStatus.UP:
-                print("ERROR - VM {0} still not up after {1} retries".format(node, args.num_of_iterations), file=sys.stderr)
-                sys.exit(-1)
-            else:
-                vm_dict[node] = vm_service
+            to_create.append(node)
+
+    # Create the VM in "blocks"
+    for block in chunks(to_create, args.block_size):
+        block_futures = []
+        for node in block:
+            vm_future = vms_service.add(types.Vm(name=node,
+                                                 cluster=types.Cluster(name=args.ovirt_cluster),
+                                                 template=types.Template(name=args.ovirt_template)), wait=False)
+            block_futures.append((node, vm_future))
+        # wait for all the VMs from this block to be created
+        for node_name, future_vm in block_futures:
+            vm = future_vm.wait()
+            vm_dict[node_name] = vms_service.vm_service(vm.id)
+        # sleep before the next block
+        time.sleep(args.sleep_between_iterations)
+
+    # Start each VM when it's created, but try to batch the calls
+    counter = 1
+    started = set()
+    pub_sshkey = os.environ[args.pub_sshkey]
+    # number of attempts is bigger here because it's not attempts per VM
+    # like in the other nodes.
+    while counter < args.num_of_iterations * len(cluster_nodes):
+        start_futures = []
+        for node_name, vm_service in vm_dict.items():
+            if node_name in started:
+                continue
+            vm = vm_service.get()
+            print("%s: vm.status = %s" % (node_name, vm.status), file=sys.stderr)
+            if vm.status == types.VmStatus.DOWN:
+                print("%s: starting" % (node_name), file=sys.stderr)
+                future = vm_service.start(use_cloud_init=True, wait=False,
+                                          vm=types.Vm(initialization=types.Initialization(authorized_ssh_keys=pub_sshkey)))
+                start_futures.append(future)
+                started.add(node_name)
+            elif vm.status == types.VmStatus.UP:
+                # make sure we don't wait forever for VMs to be down when they're
+                # already up.
+                started.add(node_name)
+
+        # wait for this batch of VMs
+        print("batch size = %s" % len(start_futures))
+        for future in start_futures:
+            future.wait()
+
+        if len(started) == len(cluster_nodes):
+            # We called .start() on all VMs
+            break
+
+        time.sleep(args.sleep_between_iterations)
+        counter += 1
+
+    if counter >= args.num_of_iterations and len(started) != len(cluster_nodes):
```

Comment:
You can use `else:` clause on a while/for loop, it's run exactly when it finishes without `break`.

Reply:
Oh, nice! I forgot that's possible. I think I'll do that + a comment with explanation, because the `else:` clause for a loop is a bit uncommon and might be confusing.